packages/codecs-core/src/errors.ts

```diff
@@ -0,0 +1,24 @@
+export class DecodingEmptyBufferError extends Error {
+    readonly name: string = 'DecodingEmptyBufferError';
+
+    constructor(codec: string) {
+        super(`Codec [${codec}] cannot decode empty buffers.`);
+    }
+}
+
+export class NotEnoughBytesError extends Error {
+    readonly name: string = 'NotEnoughBytesError';
+
+    constructor(codec: string, expected: bigint | number, actual: bigint | number) {
+        super(`Codec [${codec}] expected ${expected} bytes, got ${actual}.`);
+    }
+}
+
+export class ExpectedFixedSizeCodecError extends Error {
+    readonly name: string = 'ExpectedFixedSizeCodecError';
+
+    constructor(message?: string) {
+        message ??= 'Expected a fixed-size coded, got a variable-size one.';
+        super(message);
+    }
+}
```

Comment:
In general, error subclasses bum me out, because:

* they take up space
* they encourage `instanceof` checks which _always_ get us into trouble when someone has multiple copies of a library

The approach I want to try in web3.js is to have _coded errors_. That is to say that there's _one_ `Error` subclass for all of the entire library that has a `code` instance variable. That code is a Typescript const enum. Now you can do stuff like this:

```ts
} catch (e) {
    if (e instanceof Object && 'code' in e && e.code === SolanaCodedError.EXPECTED_FIXED_SIZE_CODEC) {
        /* ... */
    }
}
```

And that compiles down to basically:

```ts
catch (e) {
    if (e instanceof Object && 'code' in e && e.code === 0 /* SolanaCodedError.EXPECTED_FIXED_SIZE_CODEC */) {
        /* ... */
    }
}
```

Here's an example: https://github.com/solana-labs/solana-web3.js/blob/55e373b2e48becf999a67ce4cb236f8265987a60/packages/rpc-transport/src/json-rpc-errors.ts#L30-L42

I'm pretty sure there's a way to get it so that depending on what `code` ends up being, TypeScript refines `data` to be a particular type other than `unknown`. I can't find an example right now.

Reply:
In general, I like that solution. What I don't like about it is that each package will have to define its error code in some other package because they all need to be in the same enum. I'd much prefer each package to be responsible for their own errors much like the current `SolanaJsonRpcError` does.

Reply:
I think this warrants a deeper conversation anyway. I'll use simple `Error`s for now and add `// TODO: Coded error.` to stay consistent with the rest of the libraries.